### PR TITLE
Fix the logic of undelete subdir trees iteratively

### PR DESCRIPTION
**Pull request by clarenceli-msft**
The previous `undeleteSubDirectoryTree` could become indefinitely trapped in the root directory when iterating the disposed subdirectories, resulting in the "maximum call stack size exceeded" error (infinite loop) when rolling-back a deletion, given multiple subdirectories exists under the root.

Update the logic to make it undelete the subdirectories iteratively, and add an additional e2e test for more comprehensive coverage.

[AB#6054](https://dev.azure.com/fluidframework/internal/_workitems/edit/6054)

### CODE REVIEW

File: packages/dds/map/src/directory.ts [hunk 1]
```diff
@@ -2447,10 +2447,11 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
 	}
 
 	private undeleteSubDirectoryTree(directory: SubDirectory): void {
-		// Restore deleted subdirectory tree. This will unmark "deleted" from the subdirectories from bottom to top.
-		for (const [_, subDirectory] of this._subdirectories.entries()) {
```
Comment (scarlettjlee) [inline packages/dds/map/src/directory.ts:2451]: Whoops. This was very wrong. Sorry.

File: packages/dds/map/src/directory.ts [hunk 2]
```diff
@@ -2447,10 +2447,11 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
...
-			this.undeleteSubDirectoryTree(subDirectory);
-		}
+		// Restore deleted subdirectory tree. Need to undispose the current directory first, then get access to the iterator.
+		// This will unmark "deleted" from the subdirectories from top to bottom.
```
Comment (scarlettjlee) [inline packages/dds/map/src/directory.ts:2451]: Why does the order need to change to top to bottom? That seems unrelated to the infinite calls.

Comment (clarenceli-msft) [reply packages/dds/map/src/directory.ts:2451]: @scarlettjlee The order does not matter, but it is necessary to undispose the current level directory first before accessing the iterator, consequently the undispose order becomes top-to-bottom

**Review by scarlettjlee — APPROVED** [review]
